### PR TITLE
feat: Add central store schema and serializer (Issue #151)

### LINKED ISSUE #151: refactor: Create central store schema and serializer
## Overview

Field names and CSV coercions differ across scrapers (`zip` vs `postal_code`, `phone` vs `phone_number`, bools to strings). Each scraper reimplements serialization, risking dropped columns and inconsistent exports.

## Proposal

Create `src/shared/store_serializer.py`:
- Define canonical store schema with normalized field names
- Handle CSV/Excel-safe casting (formula injection already in ExportService)
- Provide optional-field defaults
- Per-retailer adapters map raw data → canonical schema

## Priority
🟠 **Tier 3: Next** - Enables consistent exports and change detection

## Open Question
- Do analysts need stable CSV column order? If so, lock order in serializer.

## Acceptance Criteria
- [ ] Canonical schema defined and documented
- [ ] Per-retailer adapters normalize raw data
- [ ] Exports use consistent field names
- [ ] Backward compatibility with existing data files

### PR DESCRIPTION
**Pull request by vmud**
## Summary

- Implements central Store dataclass with canonical field names
- Adds field normalization to handle retailer-specific variations
- Creates StoreSerializer for consistent CSV/Excel export with stable column ordering
- Maintains full backward compatibility with existing scrapers
- 32 comprehensive tests, all passing

## Problem Solved

Field names differ across scrapers causing inconsistencies in exports:
- **postal_code** (Target, AT&T, Bell, Telus, Walmart) vs **zip** (Verizon, T-Mobile, Cricket, Costco)
- **phone_number** (Walmart) vs **phone** (others)
- **lat/lng/lon** vs **latitude/longitude**

Each scraper was handling normalization independently (if at all), leading to:
- Different CSV column names for the same data
- Manual remapping needed for data analysis
- Duplicated serialization logic

### Store Dataclass
```python
@dataclass
class Store:
    # Required fields
    store_id: str
    name: str
    street_address: str
    city: str
    state: str
    
    # Optional fields with canonical names
    zip: Optional[str]  # Normalized from postal_code
    phone: Optional[str]  # Normalized from phone_number
    latitude: Optional[float]  # Normalized from lat
    longitude: Optional[float]  # Normalized from lng/lon
    # ...
```

### Field Normalization

Factory method `Store.from_raw()` automatically normalizes:
```python
# Input: retailer-specific format
raw_data = {
    'store_id': 'T1234',
    'name': 'Target Store',
    'postal_code': '55403',  # Non-standard
    'phone_number': '612-555-1234',  # Non-standard
}

# Output: canonical format
store = Store.from_raw(raw_data, retailer='target')
assert store.zip == '55403'  # Normalized
assert store.phone == '612-555-1234'  # Normalized
```

Mapping handled by `FIELD_ALIASES`:
```python
FIELD_ALIASES = {
    'postal_code': 'zip',
    'phone_number': 'phone',
    'address': 'street_address',
    'lat': 'latitude',
    'lng': 'longitude',
    'lon': 'longitude',
}
```

### CSV Export with Stable Ordering

`StoreSerializer` ensures consistent column order:
```python
FIELD_ORDER = [
    'store_id', 'name', 'street_address', 'city', 'state',
    'zip', 'country', 'phone', 'latitude', 'longitude',
    'url', 'hours', 'retailer', 'scraped_at'
]
```

Benefits:
- CSV files have same column order regardless of scraper
- Extra fields appended alphabetically for determinism
- Applies CSV injection sanitization automatically

### Basic Normalization
```python
from src.shared import Store

# Walmart data (uses phone_number, postal_code)
walmart_data = {
    'store_id': 'W5678',
    'name': 'Walmart Supercenter',
    'street_address': '200 Walmart Dr',
    'city': 'Bentonville',
    'state': 'AR',
    'postal_code': '72712',
    'phone_number': '479-555-5678'
}

store = Store.from_raw(walmart_data, retailer='walmart')
print(store.zip)  # '72712' (normalized)
print(store.phone)  # '479-555-5678' (normalized)
```

### CSV Export
```python
from src.shared import StoreSerializer

stores = [Store.from_raw(data, 'walmart') for data in raw_stores]
fields = StoreSerializer.get_ordered_fields(stores)  # Stable order
rows = [StoreSerializer.to_csv_row(store) for store in stores]  # Sanitized
```

### Quick Dict Normalization
```python
from src.shared import normalize_store_dict

# For legacy code that expects dicts
normalized = normalize_store_dict(raw_data, retailer='target')
```

## Backward Compatibility

- **Scrapers unchanged**: Existing scrapers continue using their current field names
- **Export service unchanged**: ExportService continues working as-is
- **Optional adoption**: Scrapers can migrate to Store incrementally
- **Dataclass benefits**: Type hints, validation, IDE autocomplete when adopted

## Test Coverage

32 comprehensive tests covering:
- ✅ Field alias mappings (postal_code→zip, phone_number→phone, etc.)
- ✅ Store dataclass creation with required/optional fields
- ✅ Store.from_raw() normalization for all field variations
- ✅ Extra field preservation for retailer-specific data
- ✅ StoreSerializer CSV/Excel export with sanitization
- ✅ Stable field ordering for consistent exports
- ✅ Integration scenarios for Target, Walmart, Verizon patterns
- ✅ CSV injection protection integration
- ✅ Coordinate normalization (lat/lng/lon variations)

All 32 tests passing.

## Files Changed

- **src/shared/store_serializer.py** (new): Store dataclass and StoreSerializer
- **tests/test_store_serializer.py** (new): Comprehensive test suite
- **src/shared/__init__.py** (modified): Export Store, StoreSerializer, normalize_store_dict

## Future Work

Individual scrapers can optionally adopt Store dataclass for:
- Better type safety with dataclasses
- Reduced boilerplate (no manual normalization)
- IDE autocomplete support
- Centralized validation

Example migration (optional):
```python
# Before
def parse_store(data):
    return {
        'store_id': data['id'],
        'postal_code': data['zip'],  # Manual mapping
        # ...
    }

# After
def parse_store(data):
    raw = {'store_id': data['id'], 'zip': data['zip'], ...}
    return Store.from_raw(raw, retailer='target')  # Auto-normalized
```

## Fixes

Closes #151

---

🤖 Generated by [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Low Risk**
> Mostly additive shared utilities (new module, docs, tests) with no changes to existing scraper/export call paths unless consumers opt in. Main risk is subtle differences in normalized field names/order affecting downstream code if adopted.
> 
> **Overview**
> Introduces a new shared `Store` dataclass (`store_serializer.py`) that normalizes retailer-specific field names (e.g., `postal_code`→`zip`, `phone_number`→`phone`, `lat/lng/lon`→`latitude/longitude`), preserves unknown keys as `extra_fields`, and validates required fields on creation.
> 
> Adds `StoreSerializer` helpers to produce CSV/Excel-ready rows with built-in formula-injection sanitization and stable column ordering (standard fields first, then extra fields sorted), plus a `normalize_store_dict` convenience API; these are exported via `src/shared/__init__.py`. Documentation and a comprehensive new test suite are included to cover normalization, ordering determinism, collision warnings, and edge cases.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit df8ada39afcfb6ed712f54aa4f668c4b1bdadf4e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Centralized store data normalization with retailer-specific field alias handling, deterministic CSV/Excel-safe exports, and CSV injection protection.

* **Documentation**
  * Added a comprehensive guide on store data schema, normalization patterns, CSV export format, sanitization, and integration guidance.

* **Tests**
  * Added extensive tests covering normalization, CSV serialization, alias handling, edge cases, and real-world retailer scenarios.

### CODE REVIEW

File: docs/store-serializer-guide.md [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+    country: str = "US"
+    phone: Optional[str] = None
+    latitude: Optional[float] = None
+    longitude: Optional[float] = None
+    url: Optional[str] = None
+    hours: Optional[str] = None
+    retailer: Optional[str] = None
+    scraped_at: Optional[str] = None
+    extra_fields: Optional[Dict[str, Any]] = None
+```
```
[the remark this answers is not in the document]
Comment (vmud) [reply docs/store-serializer-guide.md]: Fixed in commit 1fab8d0. Updated the docs snippet to use `extra_fields: Dict[str, Any] = field(default_factory=dict)` to match the actual implementation.

File: src/shared/store_serializer.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        """
+        # Get base dict
+        data = store.to_dict(include_extra=True, flatten=True)
+
+        # Convert all values to strings and sanitize
+        csv_row = {}
+        for key, value in data.items():
+            # Convert to string
+            if value is None:
+                str_value = ''
```
[the remark this answers is not in the document]
Comment (vmud) [reply src/shared/store_serializer.py]: Fixed in commit 1fab8d0. Added `for_csv` parameter to `to_dict()` to preserve None values when preparing CSV data. The `to_csv_row()` method now passes `for_csv=True` so None values are properly converted to empty strings rather than being filtered out.

File: src/shared/store_serializer.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+        Returns:
+            Ordered list of field names
+        """
+        if not stores:
+            return list(cls.FIELD_ORDER)
+
+        # Collect all fields from stores
+        all_fields = set()
+        for store in stores:
+            data = store.to_dict(include_extra=True, flatten=True)
```
[the remark this answers is not in the document]
Comment (vmud) [reply src/shared/store_serializer.py]: Fixed in commit d82341a. `get_ordered_fields()` now uses `for_csv=True` to match the field set produced by `to_csv_row()`, preventing `DictWriter` key mismatches.

File: src/shared/store_serializer.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+                store_data[key] = value
+            else:
+                extra[key] = value
+
+        # Store extra fields in extra_fields dict
+        if extra:
+            store_data['extra_fields'] = extra
+
+        # Validate required fields using shared validation
+        validation_result = validate_store_data(store_data)
```
[the remark this answers is not in the document]
Comment (vmud) [reply src/shared/store_serializer.py]: This issue was resolved by merging main into the feature branch. The merge brought in PR #215 (fae55d5) which updated `validate_store_data()` in `src/shared/utils.py` to check for all postal code field name variations including `zip`:

```python
postal_code = (store.get('zip') or store.get('postal_code') or
               store.get('zipcode') or store.get('zip_code') or
               store.get('postalcode'))
```

Validation now works correctly after the `postal_code` → `zip` normalization. Verified with tests:
- All 36 tests in `test_store_serializer.py` pass
- Manual test confirms postal_code normalization and validation work together

Fixed by merge commit a5ccb32.
